adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,285 @@
+	var extRegs openrtb_ext.ExtRegs
+	if request.Regs != nil {
+		if request.Regs.COPPA == 1 {
+			return nil, []error{&errortypes.BadInput{
+				Message: "COPPA is not supported",
+			}}
+		}
+		if err := json.Unmarshal(request.Regs.Ext, &extRegs); err == nil {
+			if extRegs.GDPR != nil && (*extRegs.GDPR == 1) {
+				return nil, []error{&errortypes.BadInput{
+					Message: "GDPR is not supported",
+				}}
+			}
+			if extRegs.USPrivacy != "" {
+				return nil, []error{&errortypes.BadInput{
+					Message: "CCPA is not supported",
+				}}
+			}
+		}
+	}
+
+	imp, err := setImps(request, requestInfo)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	request.Imp = imp
+
+	request.Source.Ext, err = setSourceExt()
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	request.Ext, err = setExt(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: getHeaders(request),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getHeaders(request *openrtb.BidRequest) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return headers
+}
+
+func setImps(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]openrtb.Imp, error) {
+	for i := 0; i < len(request.Imp); i++ {
+		imp := &request.Imp[i]
+
+		var ext unicornImpExt
+		err := json.Unmarshal(imp.Ext, &ext)
+
+		if err != nil {
+			return nil, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while decoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		var placementID string
+		if ext.Bidder.PlacementID != "" {
+			placementID = ext.Bidder.PlacementID
+		} else {
+			placementID, err = getStoredRequestImpID(imp)
+			if err != nil {
+				return nil, &errortypes.BadInput{
+					Message: fmt.Sprintf("Error get StoredRequestImpID from imp[%d]: %s", i, err),
+				}
+			}
+		}
+
+		ext.Bidder.PlacementID = placementID
+
+		imp.Ext, err = json.Marshal(ext)
+		if err != nil {
+			return nil, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while encoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		secure := int8(1)
+		imp.Secure = &secure
+		imp.TagID = placementID
+	}
+	return request.Imp, nil
+}
+
+func getStoredRequestImpID(imp *openrtb.Imp) (string, error) {
```

Comment:
The built in Go library isn't ideal for pulling back a single value from a JSON object. While this is valid and working code, please consider an abbreviated approach using the jsonparser library. Something like:

```
func getStoredRequestImpID(imp *openrtb.Imp) (string, error) {
	v, err := jsonparser.GetString(imp.Ext, "prebid", "storedrequest", "id")

	if err != nil && err != jsonparser.KeyPathNotFoundError {
		return "", fmt.Errorf("stored request id not found: %s", err)
	}

	return v, nil
}
```

EDIT: Ah, I see you're already using jsonparser in `setExt`.

Reply:
I used above approach. fixed.